src/vsls-redux.ts

```diff
@@ -0,0 +1,232 @@
+) => (reducer: Reducer, initialStateOrEnhancer?: DeepPartial<any> | StoreEnhancer, enhancer?: StoreEnhancer): Store => {
+  const vslsRedux = new VSLSRedux();
+  const middlewareEnhancer = applyMiddleware(vslsRedux.createMiddleware());
+  let store: Store;
+  if (typeof initialStateOrEnhancer === 'object') {
+    // DeepPartial<any>
+    const composedEnhancer: any = enhancer
+      ? compose(
+          enhancer,
+          middlewareEnhancer,
+        )
+      : middlewareEnhancer;
+    store = next(reducer, <DeepPartial<any>>initialStateOrEnhancer, composedEnhancer);
+  } else {
+    // StoreEnhancer
+    const composedEnhancer: any = initialStateOrEnhancer
+      ? compose(
+          <StoreEnhancer>initialStateOrEnhancer,
+          middlewareEnhancer,
+        )
+      : middlewareEnhancer;
+    store = next(reducer, composedEnhancer);
+  }
+  vslsRedux.setStore(store);
+  return store;
+};
+
+class VSLSRedux {
+  private remoteService: IMiddlewareProvider | null = null;
+  private store: Store | null = null;
+
+  public constructor() {
+    this.init();
+  }
+
+  private async init() {
+    const vslsAPI = await vsls.getApiAsync();
+    if (vslsAPI) {
+      vslsAPI.onDidChangeSession((e: vsls.SessionChangeEvent) => {
+        const isSessionEnded = !e.session.id;
+
+        if (isSessionEnded) {
+          this.remoteService = null;
+          return;
+        }
+
+        switch (e.session.role) {
+          case vsls.Role.Guest:
+            this.remoteService = new GuestService(this.store);
+            break;
+          case vsls.Role.Host:
+            this.remoteService = new HostService(this.store);
+            break;
+          default:
+            return;
+        }
+      });
+    }
+  }
+
+  private defaultActionFilter = (action: Action) => {
+    return action.type !== SET_INITIAL_STATE_ACTION_NAME;
+  }
+
+  public createMiddleware(actionFilter = (action: Action) => true): Middleware {
+    return (): any => {
+      return (next: (action: Action) => void) => async (action: Action) => {
+        const syncAction = this.defaultActionFilter(action) && actionFilter(action);
+        if (syncAction && this.remoteService) {
+          this.remoteService.middleware()(next)(action);
+        } else {
+          next(action);
+        }
+      };
+    };
+  }
+
+  public setStore(store: Store) {
+    this.store = store;
+  }
+}
+
+/**
+ * The HostService sends the host state to any guests that request it via the getState request
+ * and will stamp the vslsId onto any incoming actions and redistribute them to all guests.
+ */
+class HostService implements IMiddlewareProvider {
+  private service: vsls.SharedService | null = null;
+  private nextActionId = 0;
+
+  public constructor(store: Store | null) {
+    this.init(store);
+  }
+
+  private async init(store: Store | null) {
+    const vslsAPI = await vsls.getApiAsync();
+    if (vslsAPI) {
+      // TODO: Namespace this service name by the calling extension's name
```

Comment:
We really need this.

Reply:
Agreed. And it shouldn't be hard to do! Definitely the next thing we should do.